// jengine_builder/lib/windows/window.js
var U = require('./../utils')
;

var Window = module.exports = U.object.extend({
  defaults: {
    windowOptions: {
      width: 1024,
      height: 768
    }
  },

  url: '/index.html',

  open: function() {
    var self = this;
    self._window = self.appjs.createWindow(self.application.urlFor(U.getValue(self.url)), self.options.windowOptions);
    self._window.on('create', function() {
      self._window.frame.show();
      self._window.frame.center();
      self._window.frame.setMenuBar(self.createMenu());
    });
    self.trigger('open');
    return self;
  },

  createMenu: function() {
    var self = this;
    self._menu = self.appjs.createMenu([
      {
        label:'&File',
        submenu:[
          {
            label:'E&xit',
            action: function(){
              self._window.close();
            }
          }
        ]
      },
      {
        label:'&Window',
        submenu:[
          {
            label:'Fullscreen',
            action:function(item) {
              self._window.frame.fullscreen();
            }
          },
          {
            label:'Minimize',
            action:function(){
              self._window.frame.minimize();
            }
          },
          {
            label:'Maximize',
            action:function(){
              self._window.frame.maximize();
            }
          },
          {
            label:''//separator
          },
          {
            label:'Restore',
            action:function(){
              self._window.frame.restore();
            }
          }
        ]
      }
    ]);
    return self._menu;
  },
});

Window.on('initialize', function() {
  this.application = this.options.application;
  this.appjs = this.application.appjs;
});
