// jengine_builder/lib/utils/options.js
var __      = require('./underscore.js'),
    extend  = require('./extend.js'),
    events  = require('./events.js')
;

var Options = module.exports = function(options, loadDefaults) {
  if (this.__super__) {
    this._options = this.__super__.extend(this.defaults);
    __.extend(this._options, options);
  } else {
    this._options = __.extend({}, this.defaults, options);
  }
};

__.extend(Options.prototype, {
  defaults: {},

  extend: function(newOptions) {
    return __.extend(this.options(), newOptions);
  },

  extendWith: function(newOptions) {
    __.extend(this._options, newOptions);
    return this;
  },

  options: function() {
    return __.clone(this._options || {});
  },

  get: function(key) {
    return this._options[key];
  },

  getSet: function(name) {
    return this.pick(this.sets[name]);
  },

  pick: function(keys) {
    var self = this,
        optionsSet = {}
    ;
    keys || (keys = []);
    __.each(keys, function(key) {
      optionsSet[key] = self.get(key);
    });
    return optionsSet
  }
});

__.extend(Options, events, {extend: extend});
__.extend(Options.prototype, events);
