// jengine_builder/lib/application.js
var windows = require('./windows.js'),
    U       = require('./utils.js'),
    path    = require('path'),
    options = require('./options.js')  
;

var Application = module.exports = U.object.extend({
  optionsObject: options.Application,
  
  defaults: {
    rootUrl: 'http://appjs/'
  },

  closeAllWindows: function() {
    U._.each(this.windows, function(window, name) {
      window.close();
    });
  },

  intro: function() {
    this.closeAllWindows();
    this.windows.intro = new windows.Intro({application: this});
    this.windows.intro.open();
    return this.windows.intro;
  },

  project: function() {
    this.closeAllWindows();
    this.windows.project = new windows.Project({application: this});
    this.windows.project.open();
    return this.windows.project;
  },

  urlFor: function(path) {
    return this.rootUrl+path;
  }
});

Application.on('initialize', function() {
  this.root = path.dirname(process.mainModule.filename);
  this.rootUrl = this.getOption('rootUrl');
  this.appjs = require('appjs');
  this.appjs.serveFilesFrom(path.join(this.root, 'content'));
  this.windows = {};
});
